core/src/main/java/software/amazon/awssdk/core/SdkRequestOverrideConfig.java

```diff
@@ -0,0 +1,211 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * Base per-request override configuration for all SDK requests.
+ */
+public abstract class SdkRequestOverrideConfig {
+
+    private final Map<String, String> additionalHeaders;
+
+    private final Map<String, List<String>> additionalQueryParameters;
+
+    private final Duration clientExecutionTimeout;
+
+    private final String appendUserAgent;
+
+    private ReadLimitInfo readLimitInfo;
+
+    protected SdkRequestOverrideConfig(Builder<?> builder) {
+        this.additionalHeaders = builder.additionalHeaders();
+        this.additionalQueryParameters = builder.additionalQueryParameters();
+        this.clientExecutionTimeout = builder.clientExecutionTimeout();
+        this.appendUserAgent = builder.appendUserAgent();
+        this.readLimitInfo = builder.readLimitInfo();
+    }
+
+    public Optional<Map<String, String>> additionalHeaders() {
+        return Optional.ofNullable(additionalHeaders);
+    }
+
+    public Optional<Map<String, List<String>>> additionalQueryParameters() {
+        return Optional.ofNullable(additionalQueryParameters);
+    }
+
+    public Optional<Duration> clientExecutionTimeout() {
+        return Optional.ofNullable(clientExecutionTimeout);
+    }
+
+    public Optional<String> appendUserAgent() {
+        return Optional.ofNullable(appendUserAgent);
+    }
+
+    public Optional<ReadLimitInfo> readLimitInfo() {
+        return Optional.ofNullable(readLimitInfo);
+    }
+
+    public abstract Builder<? extends Builder> toBuilder();
+
+    public interface Builder<B extends Builder> {
+        Map<String, String> additionalHeaders();
+
+        B additionalHeader(String name, String value);
+
+        B additionalHeaders(Map<String, String> customHeaders);
+
+        Map<String, List<String>> additionalQueryParameters();
+
+        B additionalQueryParameter(String name, String value);
+
+        B additionalQueryParameters(Map<String, List<String>> customQueryParameters);
+
+        Duration clientExecutionTimeout();
+
+        B clientExecutionTimeout(Duration clientExecutionTimeout);
+
+        String appendUserAgent();
+
+        B appendUserAgent(String appendUserAgent);
```

Comment:
It's not clear what this does. Does it need to be supported for now? If so, can we make it more specific so they know what to put in there? (eg. Library name)

Reply:
Not that I can tell. Looks like it was used primarily by the higher level clients.

Reply:
I'd vote for being more specific with the name, then. We could then format it however we want, instead of relying on them formatting it for us.

`appendUserAgent`: Add this stuff at the end of the user agent
vs
`apiName`: The name of the high-level api library used for this request

If we expect there to be multiple API names, it could be an list/set that they can add to.

Reply:
Sounds good to me. Changed this to `apiName`